File: backend/app/schemas/pet.py
from decimal import Decimal
from typing import Annotated

from pydantic import BaseModel, Field

from .enums import Species, Gender, Size, EnergyLevel, PetStatus

PetAge = Annotated[Decimal, Field(max_digits=3, decimal_places=1, ge=0)]

class PetBase(BaseModel):
    name: str
    species: Species
    breed: str
    age: PetAge
    gender: Gender
    size: Size
    energy_level: EnergyLevel
    description: str | None = None
    photo_url: str | None = None
    adoption_fee: Decimal | None = None
    category_id: int

class PetCreate(PetBase):
    pass

class PetUpdate(BaseModel):
    name: str | None = None
    species: Species | None = None
    breed: str | None = None
    age: PetAge | None = None
    gender: Gender | None = None
    size: Size | None = None
    energy_level: EnergyLevel | None = None
    description: str | None = None
    photo_url: str | None = None
    adoption_fee: Decimal | None = None
    category_id: int | None = None

class PetOut(PetBase):
    id: int
    owner_id: int
    is_approved: bool
    status: PetStatus

    model_config = {"from_attributes": True}

File: frontend/lib/features/pets/models/pet.dart
import '../../../core/network/api_endpoints.dart';

class Pet {
  const Pet({
    required this.id,
    required this.name,
    required this.species,
    required this.breed,
    required this.age,
    required this.gender,
    required this.size,
    required this.energyLevel,
    required this.status,
    required this.categoryId,
    required this.ownerId,
    required this.isApproved,
    required this.description,
    this.photoUrl,
    this.adoptionFee,
  });

  final int id;
  final String name;
  final String species;
  final String breed;
  final double age;
  final String gender;
  final String size;
  final String energyLevel;
  final String status;
  final int categoryId;
  final int ownerId;
  final bool isApproved;
  final String description;
  final String? photoUrl;
  final double? adoptionFee;

  factory Pet.fromJson(Map<String, dynamic> json) {
    return Pet(
      id: _parseInt(json['id']),
      name: json['name']?.toString() ?? '',
      species: json['species']?.toString() ?? '',
      breed: json['breed']?.toString() ?? '',
      age: double.tryParse(json['age']?.toString() ?? '') ?? 0,
      gender: json['gender']?.toString() ?? '',
      size: json['size']?.toString() ?? '',
      energyLevel: json['energy_level']?.toString() ?? '',
      status: json['status']?.toString() ?? '',
      categoryId: _parseInt(json['category_id']),
      ownerId: _parseInt(json['owner_id']),
      isApproved: json['is_approved'] as bool? ?? false,
      description: json['description']?.toString() ?? '',
      photoUrl: _resolvePhotoUrl(json['photo_url']),
      adoptionFee: json['adoption_fee'] == null
          ? null
          : double.tryParse(json['adoption_fee'].toString()),
    );
  }

  static int _parseInt(dynamic value) {
    if (value is int) {
      return value;
    }

    return int.tryParse(value?.toString() ?? '') ?? 0;
  }

  static String? _resolvePhotoUrl(dynamic value) {
    final rawUrl = value?.toString().trim();

    if (rawUrl == null || rawUrl.isEmpty) {
      return null;
    }

    if (rawUrl.startsWith('http://') || rawUrl.startsWith('https://')) {
      return rawUrl;
    }

    final normalizedPath = rawUrl.startsWith('/') ? rawUrl : '/$rawUrl';

    return '${ApiEndpoints.baseUrl}$normalizedPath';
  }
}
